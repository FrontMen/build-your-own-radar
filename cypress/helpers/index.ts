export const dataCy = (value: string) => cy.get(`[data-cy=${value}]`);
export const dataTestId = (value: string) => cy.get(`[data-testid=${value}]`);

export const hexToRgb = (hex: string) => {
  const bigint = parseInt(hex.replace('#', ''), 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
}
