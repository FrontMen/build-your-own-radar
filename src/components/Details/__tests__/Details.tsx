import React from 'react';
import { Details } from '..';
import { AllProvidersWrapper, withAllProviders } from 'test/helpers';
import { getByRole } from '@testing-library/react';
import { mount } from 'enzyme';
import { parsedTechData } from 'test/mockData';
import {
  storeCreator,
  technologiesStateBuilder,
  filtersStateBuilder,
} from 'test/builders';

const correctQuadrant = {
  order: 0,
  color: '',
};

describe('Details', () => {
  const filterState = filtersStateBuilder({
    quadrants: [{ name: 'framework', order: 0, color: '' }],
  });
  describe('when path is incorrect', () => {
    it('should redirect to not found page when quadrant is incorrect', () => {
      const techState = technologiesStateBuilder({
        initialized: true,
        loading: false,
        data: {},
      });
      const store = storeCreator({
        ...techState,
        filters: { ...filterState.filters },
      });

      // const store = storeCreator(state);
      const wrapper = mount(<Details />, {
        wrappingComponent: AllProvidersWrapper,
        wrappingComponentProps: {
          path: '/:technology/:quadIndex',
          route: `/alpine/5`,
          store,
        },
      });

      const RedirectWrapper = wrapper.find('Redirect');

      expect(RedirectWrapper).toHaveLength(1);
      expect(RedirectWrapper.prop('to')).toMatch('not-found');
    });

    it('should redirect to not found page when technology is incorrect', () => {
      const state = technologiesStateBuilder({
        initialized: true,
        loading: false,
        data: parsedTechData,
      });
      const store = storeCreator({
        ...state,
        filters: { ...filterState.filters },
      });
      const wrapper = mount(<Details />, {
        wrappingComponent: AllProvidersWrapper,
        wrappingComponentProps: {
          path: '/:technology/:quadIndex',
          route: `/invalid-technology/2`,
          store,
        },
      });

      const RedirectWrapper = wrapper.find('Redirect');

      expect(RedirectWrapper).toHaveLength(1);
      expect(RedirectWrapper.prop('to')).toMatch('not-found');
    });
  });

  describe('when path is correct', () => {
    it('should render skeleton when data is loading', () => {
      const state = technologiesStateBuilder({
        initialized: true,
        loading: true,
        data: parsedTechData,
      });
      const store = storeCreator({
        ...state,
        filters: { ...filterState.filters },
      });
      const { getByTestId } = withAllProviders(<Details />, {
        path: '/:technology/:quadIndex',
        route: `/alpine/${correctQuadrant.order}`,
        store,
      });

      expect(getByTestId('details-skeleton')).toBeTruthy();
    });

    it('should render correctly', () => {
      const techState = technologiesStateBuilder({
        initialized: true,
        loading: false,
        data: parsedTechData,
      });
      const store = storeCreator({
        ...techState,
        filters: { ...filterState.filters },
      });
      const { getByTestId, container } = withAllProviders(<Details />, {
        path: '/:technology/:quadIndex',
        route: `/Alpine/${correctQuadrant.order}`,
        store,
      });

      expect(getByTestId('details-back-link').getAttribute('href')).toBe(
        `/quadrant/${correctQuadrant.order}`,
      );
      expect(getByTestId('details-content-title')).toHaveTextContent(
        'Timeline: Alpine',
      );
      expect(getByTestId('details-timeline-container')).toHaveStyle({
        'border-left': `4px solid ${correctQuadrant.color}`,
      });

      expect(getByTestId('details-content-title')).toHaveTextContent(
        'Timeline: Alpine',
      );

      const timeLineItem = getByTestId('details-timeline-item-0');
      expect(getByRole(timeLineItem, 'heading')).toHaveTextContent('Alpine');
      expect(container).toMatchSnapshot();
    });
  });
});
