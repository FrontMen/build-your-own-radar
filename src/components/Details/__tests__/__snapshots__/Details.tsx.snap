// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details when path is correct should render correctly 1`] = `
.c4 {
  font-family: Montserrat,san-serif;
  font-weight: 700;
  line-height: 1.15;
  font-size: 1.5em;
}

.c0 {
  grid-area: main;
  background-color: #fff;
  padding: 16px;
  height: 100%;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 16px;
  padding-left: 0;
  font-family: Roboto,san-serif;
  font-size: 1em;
  font-weight: 400;
  line-height: 1.65;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2 {
  height: 2em;
  width: 2em;
}

.c3 {
  margin: 16px 0 32px;
  text-align: center;
}

.c5 {
  border-left: 4px solid;
  margin: 50px 0% 0 30%;
  padding: 50px 0 50px 20px;
  list-style: none;
  text-align: left;
}

.c5 h3 {
  font-weight: 400;
  font-size: 1.4em;
}

.c6 {
  border-bottom: 1px dashed;
  padding-bottom: 25px;
  position: relative;
}

.c6:last-of-type {
  padding-bottom: 0;
  margin-bottom: 0;
  border: none;
}

.c6:before,
.c6:after {
  position: absolute;
  display: block;
  top: 0;
}

.c6:before {
  left: -160px;
  top: 2px;
  content: 'Jan, 2020';
  text-align: right;
  min-width: 120px;
  font-weight: bold;
  font-size: 1.1em;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.c6:after {
  box-shadow: 0 0 0 4px undefined;
  left: -28px;
  background: #fff;
  border-radius: 50%;
  height: 11px;
  width: 11px;
  content: '';
  top: 5px;
}

.c6 h3 {
  font-weight: 400;
  font-size: 1.4em;
}

.c7 {
  background: #f3f3f3;
  max-width: 205px;
  padding: 10px;
  color: #37383d;
  border: 1px solid #f3f3f3;
  border-radius: 6px;
  word-break: break-word;
  font-family: Montserrat,san-serif;
  font-size: 0.9em;
  line-height: 20px;
}

@media (min-width:481px) {
  .c4 {
    font-size: 1.5em;
    margin-top: 0;
    margin-bottom: 16px;
  }
}

@media (min-width:481px) {
  .c0 {
    padding: 32px;
  }
}

<div>
  <div
    class="PageSlots__MainContentSlot-sc-1quj5xc-1 c0"
  >
    <a
      class="c1"
      data-testid="details-back-link"
      href="/quadrant/0"
    >
      <svg
        class="c2"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M216.4 163.7c5.1 5 5.1 13.3.1 18.4L155.8 243h231.3c7.1 0 12.9 5.8 12.9 13s-5.8 13-12.9 13H155.8l60.8 60.9c5 5.1 4.9 13.3-.1 18.4-5.1 5-13.2 5-18.3-.1l-82.4-83c-1.1-1.2-2-2.5-2.7-4.1-.7-1.6-1-3.3-1-5 0-3.4 1.3-6.6 3.7-9.1l82.4-83c4.9-5.2 13.1-5.3 18.2-.3z"
        />
      </svg>
      <span>
        quadrant.framework.name
      </span>
    </a>
    <div
      class="c3"
    >
      <h2
        class="c4"
        data-testid="details-content-title"
      >
        Record: Alpine
      </h2>
      <ul
        class="c5"
        color="framework"
        data-testid="details-record-container"
      >
        <li
          class="c6"
          color="framework"
          data-testid="details-record-item-0"
          value="Jan, 2020"
        >
          <h3>
            Alpine
          </h3>
          <div
            class="c7"
          >
            <span>
              details.emptyDescription
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
